midend/hsIndexSimplify.cpp

```diff
@@ -0,0 +1,127 @@
+    if (isIf = statement->is<IR::IfStatement>()) {
+        auto* curIf = statement->to<IR::IfStatement>();
+        elseBody = curIf->ifFalse;
+        curIf->condition->apply(aiFinder);
+    } else if (statement->is<IR::SwitchStatement>()) {
+        auto* curSwitch = statement->to<IR::SwitchStatement>();
+        curSwitch->expression->apply(aiFinder);
+    } else {
+        statement->apply(aiFinder);
+    }
+    if (aiFinder.arrayIndex == nullptr) {
+        // Remove concrete indexes.
+        HSIndexFinder aiElim(nullptr, 0);
+        return const_cast<IR::Node*>(statement->apply(aiElim));
+    }
+    
+    IR::IfStatement* result = nullptr;
+    IR::IfStatement* curResult = nullptr;
+    size_t sz = aiFinder.getArraySize();
+    for (size_t i = 0; i < sz; i++) {
+        HSIndexFinder aiRewriter(aiFinder.arrayIndex, i);
+        IR::IfStatement* newIf = nullptr;
+        auto* newStatement = statement->apply(aiRewriter)->to<IR::Statement>();
+        auto* newIndex = aiFinder.arrayIndex->right->apply(aiRewriter)->to<IR::Expression>();
+        if (isIf) {
+            newIf = newStatement->to<IR::IfStatement>()->clone();
+            newIf->condition = new IR::LAnd(
+                new IR::Equ(newIndex, new IR::Constant(aiFinder.arrayIndex->right->type, i)),
+                newIf->condition);
+        } else {
+            newIf = new IR::IfStatement(
+                new IR::Equ(newIndex, new IR::Constant(aiFinder.arrayIndex->right->type, i)),
+                newStatement, nullptr);
+        }
+        if (result == nullptr) {
+            result = newIf;
+            curResult = newIf;
+        } else {
+            curResult->ifFalse = newIf;
+            curResult = newIf;
+        }
+    }
+    curResult->ifFalse = elseBody;
+    return result;
+}
+
+IR::Node* HSIndexSimplifier::preorder(IR::AssignmentStatement* assignmentStatement) {
+    return eliminateArrayIndexes(assignmentStatement);
+}
+
+IR::Node* HSIndexSimplifier::preorder(IR::ConstructorCallExpression* expr) {
+    // Eliminate concrete indexes.
+    auto* newExpr = expr->clone();
+    auto* newArguments = new IR::Vector<IR::Argument>();
+    HSIndexFinder aiRewriter(nullptr, 0);
+    for (const auto* arg : *expr->arguments)
+        newArguments->push_back(arg->apply(aiRewriter)->to<IR::Argument>());
+    newExpr->arguments = newArguments;
+    return newExpr;
+}
+
+IR::Node* HSIndexSimplifier::preorder(IR::IfStatement* ifStatement) {
+    return eliminateArrayIndexes(ifStatement);
+}
+
+IR::Node* HSIndexSimplifier::preorder(IR::MethodCallStatement* methodCallStatement) {
+    auto* newMethodCall = methodCallStatement->clone();
+    HSIndexFinder aiRewriter(nullptr, 0);
+    newMethodCall->methodCall =
+        newMethodCall->methodCall->apply(aiRewriter)->to<IR::MethodCallExpression>();
+    return newMethodCall;
+}
+
+IR::Node* HSIndexSimplifier::preorder(IR::SelectExpression* selectExpression) {
+    // All non-concrete indexes should be eliminated by ParserUnroll.
```

Comment:
Is this a hard requirement? If that is not the case, a BUG should be thrown here.

Reply:
Or simply do nothing and do not convert the index here.